qp_klp/SequencingTech.py

```diff
@@ -0,0 +1,256 @@
+        def get_config(command):
+            try:
+                return self.pipeline.get_software_configuration(command)
+            except PipelineError as e:
+                if str(e) != f"'{command}' is not defined in configuration":
+                    # re-raise the Error if it's not the one we are expecting
+                    # and can properly handle here.
+                    raise PipelineError(e)
+
+        # The choice to use either bcl2fastq or bcl-convert is set in the
+        # configuration file that matches the parameters of the sequencing
+        # run found in the user input file. The Pipeline object is smart
+        # enough to select the right configuration file based on the
+        # sample sheet passed to it.
+
+        # However in this mixin, we don't know which configuration file we
+        # will get and thus we must check for the presence or absence of
+        # both dictionaries in the configuration file at run-time and make
+        # a determination based on that.
+        bcl2fastq_conf = get_config('bcl2fastq')
+        bclcnvrt_conf = get_config('bcl-convert')
+
+        if bclcnvrt_conf is None and bcl2fastq_conf is None:
+            raise PipelineError("bcl-convert and bcl2fastq sections not "
+                                "defined in configuation profile.")
+
+        # if both are defined, we will use bcl-convert by default.
+        config = bcl2fastq_conf if bclcnvrt_conf is None else bclcnvrt_conf
+
+        job = ConvertJob(self.pipeline.run_dir,
+                         self.pipeline.output_path,
+                         # for amplicon runs, get_sample_sheet_path() will
+                         # return the path for a generated dummy sheet.
+                         self.pipeline.get_sample_sheet_path(),
+                         config['queue'],
+                         config['nodes'],
+                         config['nprocs'],
+                         config['wallclock_time_in_minutes'],
+                         config['per_process_memory_limit'],
+                         config['executable_path'],
+                         config['modules_to_load'],
+                         self.master_qiita_job_id)
+
+        self.raw_fastq_files_path = join(self.pipeline.output_path,
+                                         'ConvertJob')
+
+        job.run(callback=self.status_update_callback)
+
+        # audit the results to determine which samples failed to convert
+        # properly. Append these to the failed-samples report and also
+        # return the list directly to the caller.
+        failed_samples = job.audit(self.pipeline.get_sample_ids())
+        if hasattr(self, 'fsr'):
+            # NB 16S does not require a failed samples report and
+            # it is not performed by SPP.
+            self.fsr.write(failed_samples, job.__class__.__name__)
+
+        return failed_samples
+
+
+class TellSeq(SequencingTech):
+    seqtech_type = SEQTECH_NAME_TELLSEQ
```

Comment:
I'm not sure I understand the class organization. Remember that TellSeq is a protocol which uses Illumina instruments

Reply:
What I have is an observed need to be able to mix in either the legacy method to convert bcl files to fastq or the tellseq-based method to do the same; that's all that the classes are currently doing. I'm happy to go with whatever naming choices you suggest. Perhaps SequencingTech should be renamed to Protocol? Then TellSeq can keep its name and Illumina is renamed to Standard?